rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,295 @@
+# SlateDB Timestamps & Time-To-Live
+
+<!-- TOC start (generated with https://github.com/derlin/bitdowntoc) -->
+
+- [Motivation](#motivation)
+- [Goals](#goals)
+- [Out of Scope](#out-of-scope)
+- [Public API](#public-api)
+  * [Configuration](#configuration)
+  * [Gets](#slatedbget)
+- [Implementation](#implementation)
+  * [Data Format](#data-format)
+  * [Compaction](#compaction)
+
+<!-- TOC end -->
+
+Status: In Discussion
+
+Authors:
+* [Almog Gavra](https://github.com/agavra)
+
+References:
+* https://github.com/slatedb/slatedb/issues/218
+* https://github.com/slatedb/slatedb/issues/225
+* https://github.com/facebook/rocksdb/wiki/User-defined-Timestamp
+* https://github.com/facebook/rocksdb/wiki/Time-to-Live
+* https://github.com/facebook/rocksdb/wiki/Compaction-Filter
+* https://dl.acm.org/doi/pdf/10.1145/3483840
+* https://disc-projects.bu.edu/lethe/
+* https://dl.acm.org/doi/10.1145/3555041.3589719
+
+## Motivation
+
+SlateDB does not have a native understanding of time in data operations (it does
+in regard to compaction/garbage collection, discussion on this later). This introduces
+additional complexity to operators that want to implement semantics that depend on the
+insertion/modification time of rows. The most common usages of timestamps are:
+
+1. **Time-To-Live**: many applications, either for compliance or space-saving purposes
+   require purging data a certain amount of time after its insertion. Use cases with time
+   series data and privacy requirements such as GDPR come to mind here, though ensuring that 
+   deleted data is collected within a time bound is left for a future exercise.
+2. **Resolving Out-of-Order Operations**: some applications may process incoming requests
+   in a manner that results in out-of-order insertions to SlateDB (stream processing applications
+   are frequently subject to this type of pattern). Using insertion timestamps to resolve which
+   version of a key is more recent may be preferable to using the clock at the time of insertion.
```

Comment:
This goes beyond what RocksDB supports, we can consider omitting it from the scope of the design. RocksDB enforces that users submit timestamps that are monotonically increasing.

-------- 

Quoted from: https://github.com/facebook/rocksdb/wiki/User-defined-Timestamp

Timestamp-sequence ordering constraint

For any two internal keys of the same user key, (key, ts1, seq1) and (key, ts2, seq2)

    If seq1 < seq2, then ts1 <= ts2
    If ts1 < ts2, then seq1 < seq2

Reply:
I like the idea of supporting stream time (and other timestamps) with user-supplied clocks. I do need to think a bit about/better understand the implications of this, though...